docs/proposals/service-discovery.md

```diff
@@ -0,0 +1,102 @@
+
+## Endpoint URL, Service Type
+
+A URL can accurately describe the location of a Service. A generic URL is of the following form
+
+scheme:[//[user:password@]host[:port]][/]path[?query][#fragment]
+
+however for the purpose of service discovery we can simplify this to the following form
+
+scheme:[//host[:port]][/]path
+
+If a user and/or password is required then this information can be passed using Kubernetes Secrets. Kubernetes contains the host and port of each service but it lacks the scheme and path.
+
+Service Path - Every Service has one (or more) endpoint. As a rule the endpoint should be located at the root "/" of the localtion URL, i.e. http://172.100.1.52/. There are cases where this is not possible and the actual service endpoint could be located at http://172.100.1.52/cxfcdi. The Kubernetes metadata for a service does not capture the path part, making it hard to consume this service.
+
+Service Scheme - Services can be deployed using different schemes. Some popular schemes include 'http','https','file','ftp' and 'jdbc'.
+
+Service Type - Services use different protocols that clients need to speak in order to communicate with the service, some examples of service level protocols are SOAP, REST (Yes, technically REST isn’t a protocol but an architectural style). For service consumers it can be hard to tell what protocol is expected.
+
+## Service Description
+
+The API of a service is the point of interaction with a service consumer. The description of the API is an essential piece of information at creation time of the service consumer. It has become common to publish a service definition document on a know location on the service itself. This 'well known' place it not very standard, so it is proposed the service developer provides the service description path and the type of Definition Language (DL) used.
+
+Service Description Path - To facilitate the consumption of the service by client, the location this document would be greatly helpful to the service consumer. In some cases the client side code can be generated from such a document. It is assumed that the service description document is published somewhere on the service endpoint itself.
+
+Service Description Type - A number of Definition Languages (DL) have been developed to describe the service. Some of examples are WSDL, WADL and Swagger. In order to consume a decription document it is good to know the type of DL used.
+
+## Standard Service Annotations
+
+Kubernetes allows the creation of Service Annotations. Here we propose the use of the following standard annotations
+
+'apiman.io/servicepath' - the path part of the service endpoint url. An example value could be 'cxfcdi',
+
+'apiman.io/servicescheme' - the scheme part of the service endpoint url. An example values could be 'http' or 'https'.
+
+'apiman.io/servicetype' - the protocol of the service. Example values can be 'SOAP' or 'REST',
```

Comment:
I can live with either proposal, but "arch" is not a great abbreviation, so I prefer api-protocol